src/main.zig

```diff
@@ -74,27 +75,39 @@ pub fn main() !u8 {
 
     // Load all provided executables:
     for (cli.positionals) |file_path| {
-        var elf_file = try std.fs.cwd().openFile(file_path, .{});
-        defer elf_file.close();
-
-        var source = std.io.StreamSource{ .file = elf_file };
-        var header = try std.elf.Header.read(&source);
-
-        var pheaders = header.program_header_iterator(&source);
-        while (try pheaders.next()) |phdr| {
-            if (phdr.p_type != std.elf.PT_LOAD)
-                continue; // Header isn't lodead
-
-            const dest_mem = if (phdr.p_vaddr >= 0x0080_0000)
-                &sram.data
-            else
-                &flash_storage.data;
-
-            const addr_masked: u24 = @intCast(phdr.p_vaddr & 0x007F_FFFF);
-
-            try source.seekTo(phdr.p_offset);
-            try source.reader().readNoEof(dest_mem[addr_masked..][0..phdr.p_filesz]);
-            @memset(dest_mem[addr_masked + phdr.p_filesz ..][0 .. phdr.p_memsz - phdr.p_filesz], 0);
+        var file = try std.fs.cwd().openFile(file_path, .{});
+        defer file.close();
+
+        switch (cli.options.format) {
+            .elf => {
+                var source = std.io.StreamSource{ .file = file };
+                var header = try std.elf.Header.read(&source);
+
+                var pheaders = header.program_header_iterator(&source);
+                while (try pheaders.next()) |phdr| {
+                    if (phdr.p_type != std.elf.PT_LOAD)
+                        continue; // Header isn't lodead
+
+                    const dest_mem = if (phdr.p_vaddr >= 0x0080_0000)
+                        &sram.data
+                    else
+                        &flash_storage.data;
+
+                    const addr_masked: u24 = @intCast(phdr.p_vaddr & 0x007F_FFFF);
+
+                    try source.seekTo(phdr.p_offset);
+                    try source.reader().readNoEof(dest_mem[addr_masked..][0..phdr.p_filesz]);
+                    @memset(dest_mem[addr_masked + phdr.p_filesz ..][0 .. phdr.p_memsz - phdr.p_filesz], 0);
+                }
+            },
+            .binary, .bin => {
+                const size = try file.readAll(&flash_storage.data);
+                std.debug.assert(size <= flash_storage.data.len);
```

Comment:
I guess readAll will stop once it hits the end of data, so this will always pass? I could seek_end to find the file size and assert there?

Reply:
Yeah, I want to make sure that we warn the user if they are trying to read a flash binary that is bigger than the chip can fit.